File: io/lucamienert/basic/ast/BinaryExpression.java
package io.lucamienert.basic.ast;

import io.lucamienert.basic.instruction.Instruction;
import io.lucamienert.basic.instruction.InstructionSequence;
import io.lucamienert.basic.instruction.OpCode;

public class BinaryExpression extends Expression {

    private final BinaryOperator operator;
    private final Expression leftExpression;
    private final Expression rightExpression;

    public BinaryExpression(BinaryOperator operator, Expression leftExpression, Expression rightExpression) {
        this.operator = operator;
        this.leftExpression = leftExpression;
        this.rightExpression = rightExpression;
    }

    public BinaryOperator getOperator() {
        return operator;
    }

    public Expression getLeftExpression() {
        return leftExpression;
    }

    public Expression getRightExpression() {
        return rightExpression;
    }

    @Override
    public void compile(InstructionSequence seq) {
        leftExpression.compile(seq);
        rightExpression.compile(seq);
        switch (operator) {
            case PLUS:
                seq.append(new Instruction(OpCode.ADD));
                break;
            case MINUS:
                seq.append(new Instruction(OpCode.SUB));
                break;
            case MULT:
                seq.append(new Instruction(OpCode.MUL));
                break;
            case DIV:
                seq.append(new Instruction(OpCode.DIV));
                break;
            default:
                throw new AssertionError();
        }
    }
}

File: src/BasicCompiler/Ast/BinaryExpression.cs
using BasicCompiler.Instructions;

namespace BasicCompiler.Ast;

public class BinaryExpression : Expression
{
    public BinaryOperator Operator { get; set; }
    public Expression LeftExpression { get; set; }
    public Expression RightExpression { get; set; }

    public BinaryExpression(BinaryOperator op, Expression leftExpression, Expression rightExpression)
    {
        Operator = op;
        LeftExpression = leftExpression;
        RightExpression = rightExpression;
    }

    public override void Compile(InstructionSequence instructionSequence)
    {
        LeftExpression.Compile(instructionSequence);
        RightExpression.Compile(instructionSequence);

        switch(Operator)
        {
            case BinaryOperator.PLUS: instructionSequence.Append(new Instruction(OpCode.ADD)); break;
            case BinaryOperator.MINUS: instructionSequence.Append(new Instruction(OpCode.SUB)); break;
            case BinaryOperator.MULT: instructionSequence.Append(new Instruction(OpCode.MUL)); break;
            case BinaryOperator.DIV: instructionSequence.Append(new Instruction(OpCode.DIV)); break;
            default: throw new AssertionException(Operator);
        }
    }
}
